src/sss_client/common.c

```diff
@@ -76,6 +78,12 @@ __attribute__((destructor))
 #endif
 void sss_cli_close_socket(void)
 {
+#ifdef HAVE_PTHREAD_EXT
+    if (sss_sd_key_initialized_status) {
+        pthread_key_delete(sss_sd_key);
+        sss_sd_key_initialized_status = false;
```

Comment:
Hi,

maybe `sss_sd_key_initialized_status = false;` can be called first to minimize the risk for race conditions?

I wonder if `pthread_once()` can be used here as well for the de-initialization?

bye,
Sumit

Reply:
> I wonder if pthread_once() can be used here as well for the de-initialization?

Good idea.
Another thing is that `sss_cli_close_socket()` isn't merely a lib destructor but called in multiple places...

Reply:
Actually, if a key is deleted in a lib destructor (that is only called at `dlclose()`) then `pthread_once()` isn't needed.

But we still need a flag to indicate key was actually created (or move key creation to lib constructor).

Reply:
We wouldn't need this bool flag if it would be possible to read status of `pthread_once_t` variable, but I didn't find a portable way so far.